modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,357 @@
+
+== Composite Types
+
+Composite types contains other values inside it. These composite types are Array, Object and Function.
+
+=== Array Type
+
+An Array is of type `Array&lt;T&gt;` where `T` is a type parameter that defines the type of the elements inside the Array. For example, the syntax to define a variable with the type Array of `Number` is:
+
+----
+var idsList: Array<Number> = [1, 22, 333, 4444]
+----
+
+The variable `idsList` only passes the type checking phase when the array assigned to it only contains Numbers.
+
+=== Object Type
+
+In order to define an object type there are two ways, either `Object` or `{}`, they both define an open object and don't specify any constraint to their key value pairs. Open and close objects are later defined in this section.
+
+The syntax to define an object type with a set of constraints for its key value pairs, is very similar to defining an object value:
+
+----
+{
+    keyName @(attrName: AttrType): valueType,
+    ...
+}
+----
+
+For example, to define a user type with `firstName` String, `lastName` String and `age` Number, the syntax is:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+type User = {
+    firstName: String,
+    lastName: String,
+    age: Number
+}
+----
+
+You can apply some modifiers to each key value pair to determine if a field is always present or if the field can be repeated or not.
+
+For repeated fields use `*`. For example, the following type allows to have multiple `lastName` fields:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age: Number
+}
+----
+
+For conditional fields use `?`. For example, the following type allows the age field to be present or not:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age?: Number
+}
+----
+
+==== Close and Open Objects
+
+Close object means that the type only accepts an object value if there are no additional fields in the object,
+just those key values specified by the type.
+
+Open object only put constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accepts objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.
+
+The following example works even if `age` is not mentioned in the type declaration because the type of the variable `user` is an open object:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+var user: {firstName: String, lastName: String} = {firstName: "John", lastName: "Smith", age: 34}
+----
+
+The syntax to specify a close object is `{| |}`. The following example, makes our user type only support the keys `firstName` and `lastName`, so this throws an exception because the `age` field is not accepted.
```

Comment:
```suggestion
The syntax to specify a closed object is `{| |}`. The following example makes the `User` type support only the keys `firstName` and `lastName`. The script throws an exception because the `age` field is not accepted.
```